File: backend/app/api/controllers/instructor.py
from flask import Blueprint, g, request, jsonify

from backend.app.api.dtos.auth_middleware import require_auth
from backend.app.api.dtos.response import ResponseContract
from backend.app.api.dtos.validation import error_response, validate
import backend.app.application.services.instructor as svc
import backend.app.application.services.audit as audit

instructor_bp = Blueprint("instructors", __name__, url_prefix="/api/instructors")


@instructor_bp.route("", methods=["GET"])
@require_auth
def list_instructors():
    try:
        page      = int(request.args.get("page", 1))
        page_size = int(request.args.get("page_size", 20))
        search    = request.args.get("search", "").strip() or None
        data, total = svc.list_instructors(page, page_size, search)
        return jsonify(ResponseContract(True, "OK", data, total=total).to_dict()), 200
    except Exception as e:
        return error_response(e)


@instructor_bp.route("/<instructor_id>", methods=["GET"])
@require_auth
def get_instructor(instructor_id):
    try:
        data = svc.get_instructor_by_id(instructor_id)
        if not data:
            return jsonify(ResponseContract(False, "Instructor not found.").to_dict()), 404
        return jsonify(ResponseContract(True, "OK", data[0]).to_dict()), 200
    except Exception as e:
        return error_response(e)


@instructor_bp.route("", methods=["POST"])
@require_auth
def create_instructor():
    try:
        body = request.get_json()
        validate(body, "instructor")
        result = svc.create_instructor(body)
        audit.log(g.user.id, "CREATE", "instructor",
                  result[0].get("instructor_id") if result else None, None, body)
        return jsonify(ResponseContract(True, "Instructor created.", result).to_dict()), 201
    except Exception as e:
        return error_response(e)


@instructor_bp.route("/<instructor_id>", methods=["PUT"])
@require_auth
def update_instructor(instructor_id):
    try:
        body = request.get_json()
        validate(body, "instructor", partial=True)
        result = svc.update_instructor(instructor_id, body)
        audit.log(g.user.id, "UPDATE", "instructor", instructor_id, None, body)
        return jsonify(ResponseContract(True, "Instructor updated.", result).to_dict()), 200
    except Exception as e:
        return error_response(e)


@instructor_bp.route("/<instructor_id>", methods=["DELETE"])
@require_auth
def delete_instructor(instructor_id):
    try:
        svc.delete_instructor(instructor_id)
        audit.log(g.user.id, "DELETE", "instructor", instructor_id)
        return jsonify(ResponseContract(True, "Instructor deleted.").to_dict()), 200
    except Exception as e:
        return error_response(e)

File: frontend/src/features/instructors/api/instructor.ts
import { apiFetch, apiJsonPaged } from "@/lib/api";

export type TeachingRestriction = {
    requirement_type: "min_student_age" | "max_student_age";
    value: string;
};

export type Instructor = {
    instructor_id: string;
    name: string;
    email: string | null;
    phone: string | null;
    restrictions?: TeachingRestriction[];
};

export async function getInstructors(): Promise<Instructor[]> {
    const res = await apiFetch("/api/instructors");
    if (!res.ok) throw new Error(`Failed to fetch instructors: ${res.statusText}`);
    const body = await res.json();
    return body.data ?? [];
}

export async function listInstructors(
    page = 1,
    pageSize = 20,
    search?: string
): Promise<{ data: Instructor[]; total: number }> {
    const params = new URLSearchParams({ page: String(page), page_size: String(pageSize) });
    if (search) params.set("search", search);
    return apiJsonPaged<Instructor>(`/api/instructors?${params}`);
}

export async function createInstructor(data: { name: string; email?: string; phone?: string; restrictions?: TeachingRestriction[] }): Promise<Instructor> {
    const res = await apiFetch("/api/instructors", {
        method: "POST",
        body: JSON.stringify(data),
    });
    if (!res.ok) throw new Error(`Failed to create instructor: ${res.statusText}`);
    const body = await res.json();
    return body.data;
}

export async function updateInstructor(instructorId: string, data: { name?: string; email?: string; phone?: string; restrictions?: TeachingRestriction[] }): Promise<Instructor> {
    const res = await apiFetch(`/api/instructors/${instructorId}`, {
        method: "PUT",
        body: JSON.stringify(data),
    });
    if (!res.ok) throw new Error(`Failed to update instructor: ${res.statusText}`);
    const body = await res.json();
    return body.data;
}

export async function deleteInstructor(instructorId: string): Promise<void> {
    const res = await apiFetch(`/api/instructors/${instructorId}`, { method: "DELETE" });
    if (!res.ok) throw new Error(`Failed to delete instructor: ${res.statusText}`);
}
